### PR TITLE
Added MSB 12 to 16 bit conversion

### PR DESCRIPTION
**Pull request by ShaquilleLouisa-LambertInstruments**
Currently the LiFlimReader does not support reading msb(most significant bit) 12 to 16 bit conversion.
I have added the functionality to check if msb or lsb should be used based on the packing value, if the packing value equals msb the new msb 12 to 16 bit conversion is used.

If there are any questions please contact me: Shaquile@LambertInstruments.com

### CODE REVIEW

**Comment by dgault** [comment]
Hi @ShaquilleLouisa-LambertInstruments, thank you for contributing this PR to the project. We have introduced a Contributor License Agreement, would you be able to sign and return the form following the instructions on https://ome-contributing.readthedocs.io/en/latest/cla.html

Also do you have any suitable sample files which could be used to test this PR with? If you need a suitable upload location for sample files then we recommend using [Zenodo](https://www.google.com/search?q=zenodo&oq=zeno&aqs=chrome.1.69i57j69i59j46i433i512j46i175i199i512j0i433i512j46i131i175i199i433i512j69i65l2.2345j0j7&sourceid=chrome&ie=UTF-8)

**Comment by ShaquilleLouisa-LambertInstruments** [comment]
This PR can be tested with the following files:
https://drive.google.com/file/d/1VWsavXPDpeA5JkMKatKmbtVaFHakkqE6/view?usp=share_link
https://drive.google.com/file/d/1iOyms_e0NWJpmAYIOFc6UN24OE0dOWnJ/view?usp=share_link

The first file is a recording with msb packing and the second file is a recording with lsb packing. The msb 12 to 16 bit conversion has been added to support opening the first file in fiji using bioformats. This can be tested by opening both files in fiji and confirming that the images look similar. The sample recordings are very dark, but that's not a problem. Without this additional conversion this wasn't possible, the opening of a msb packed recording resulted in vertical white lines in the image.

**Comment by ShaquilleLouisa-LambertInstruments** [comment]
Hi @dgault Thank you for looking into this PR, I have signed the Contributor License Agreement and shared the sample files in my previous comment.

**Comment by dgault** [comment]
Thanks @ShaquilleLouisa-LambertInstruments, I have marked the PR for inclusion in our CI and nightly repository tests. If they remain green we will aim to get this PR reviewed in the next few days.

File: components/formats-gpl/src/loci/formats/in/LiFlimReader.java [hunk 1]
```diff
@@ -667,25 +673,39 @@ else if (UINT12streamSeries == 1 && DarkImage != null && DarkImage.equals(new St
...
+    byte[] image16 = new byte[image.length * 4 / 3];
+
+    if (image16.length / 4 != image.length / 3)
+      return new byte[0];
+
+    for (int idx = 0, idx16 = 0; idx < (image.length - 2) && (idx16 < image16.length - 3); idx += 3, idx16 += 4) {
+      image16[idx16] = (byte)(image[idx]);
+      image16[idx16 + 1] = (byte)((image[idx + 1] & 0x0f));
+      image16[idx16 + 2] = (byte)(((image[idx + 1]) >> 4) | ((image[idx + 2]) << 4));
+      image16[idx16 + 3] = (byte)((image[idx + 2]) >> 4);
```
Comment (dgault) [inline components/formats-gpl/src/loci/formats/in/LiFlimReader.java:687]:
```suggestion
      image16[idx16] = (byte)(image[idx] & 0xff);
      image16[idx16 + 1] = (byte)((image[idx + 1] & 0x0f));
      image16[idx16 + 2] = (byte)(((image[idx + 1] & 0xf0) >> 4) | ((image[idx + 2] & 0x0f) << 4));
      image16[idx16 + 3] = (byte)((image[idx + 2] & 0xf0) >> 4);
```

**Review by dgault — CHANGES_REQUESTED** [review]
Hi @ShaquilleLouisa-LambertInstruments, thank you again for submitting the contribution. 

Overall the addition of support for MSB packing looks solid, I noticed when testing that some of the pixels values for the LSB sample had been unexpectedly changed. I believe this was due to bit operations being carried out on signed bytes, I have added some suggested changes which should resolve this. 

Also, the sample files provided are using a `flc` extension while currently the reader only recognises `fli`. If you want to add support for the flc extension also then the constructor can be updated to use the below:
`super("LI-FLIM", new String[] {"fli", "flc"});`

**Comment by ShaquilleLouisa-LambertInstruments** [comment]
Hi @dgault, thank you for reviewing my PR.

I have commited the suggested changes, but I'm a bit confused as to why the files that I shared did not have the fli extension. Are you sure they are not .fli files?

**Comment by dgault** [comment]
Just catching up with this PR again after the holidays, thanks you for the extra commits. I checked the link for the sample files again, indeed they do appear to be using the fli extension, though when I click download it automatically saves them as flc for some reason. Simply renaming the extension works as expected though.

**Comment by ShaquilleLouisa-LambertInstruments** [comment]
Hi @dgault, since there are no longer any problems with this PR and all automatic checks have been passed 
can my PR now be merged?

**Comment by dgault** [comment]
Hi @ShaquilleLouisa-LambertInstruments, I had re-tested the PR last week and was happy with the results. I then configured the sample files you provided and added them to our nightly repository tests. Those tests then failed due to the files not being closed properly and open file handles detected. (https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-test-folder/224415/console). This isn't related to your changes but was an issue with the existing code, it seems this was the first time it had been tested as part of our CI with 12 bit packed data. I have made a suggested code change which should resolve the issue and close the file handles as expected. Would you be OK with also including this change as part of this PR?

**Comment by ShaquilleLouisa-LambertInstruments** [comment]
Hi @dgault, Thank you for looking into my PR again. I don't mind adding the requested code change you mentioned but I don't see any new requested code changes.

**Comment by dgault** [comment]
Thanks @ShaquilleLouisa-LambertInstruments, with that one last suggested change we should be good to get this merged.

**Comment by ShaquilleLouisa-LambertInstruments** [comment]
Sounds good, can this now get merged?

**Comment by dgault** [comment]
All the builds and test are now green with this PR included, looks good to merge.

**Pull request merged**